flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java

```diff
@@ -282,4 +282,21 @@ public enum VertexDescriptionMode {
                     .withDescription(
                             "Whether name of vertex includes topological index or not. "
                                     + "When it is true, the name will have a prefix of index of the vertex, like '[vertex-0]Source: source'. It is false by default");
+
+    @PublicEvolving
+    public static final ConfigOption<Boolean> ALLOW_UNALIGNED_SOURCE_SPLITS =
+            key("pipeline.watermark-alignment.allow-unaligned-source-splits")
+                    .booleanType()
+                    .defaultValue(false)
```

Comment:
Please check the dev mailing list discussion. We want to change the the old behaviour from FLINK-1.14/1.15. In a couple of releases (1.18? 1.19?), we will want to disallow partially working watermark alignment when source doesn't support per split alignment. At the same time we want to provide users time to adjust/migrate if they are relaying on this partially working feature for now.

Reply:
I've read the mailing list thread. I still think it would be better to not enforce this feature for now. If I understand correctly, watermark alignment without split alignment still works but it is possible for splits to overtake each other which can slow down the processing due to regular watermark alignment. I would take it to the mailing list if we want to revisit this decision.

Reply:
> watermark alignment without split alignment still works but it is possible for splits to overtake each other which can slow down the processing due to regular watermark alignment

I'm not sure if I understand you here. So let me rephrase the consensus.

On the current master. Watermark alignment works to a small extent if a source operator has more than one split, but it doesn't work fully. In worst case, it won't work at all. Imagine if the slowest split, that is a split that's the most lagging/blocking watermark emission, is processed on the same operator that has the fastest split. In such scenario, watermark alignment without per split alignment, can not block processing of this whole operator (or otherwise it would deadlock the watermarks emission), so the fastest split will continue producing records, for example blowing up time windowed stated downstream. Even in average case, such watermark alignment will work very badly. 

Moreover, using pre-FLIP-217 watermark alignment with multiple splits per operator has been always discouraged. With FLIP-217 the consensus was, that it would be actually hurtful for the users, to have _SILENTLY_ hit such scenarios: user could configure watermark alignment, but actually it wouldn't work. The consensus was that from now on this feature should either work fully, or not at all if users tries to enable it. But we also want to provide gradual migration path for any users that would be affected by this change. Instead of forcing user to both upgrade Flink version AND solve the lack of per-split alignment on his side, he can temporarily set `allow-unaligned-source-splits` to `true`, do the Flink upgrade, and then in the background prepare for next Flink release that will remove this option and set it to `false` in stone.

Reply:
I fully agree. I've left a suggestion in this regard.